data/campaigns/fri05.wmf/scripting/mission_thread.lua

```diff
@@ -0,0 +1,236 @@
+         for _,proposal in ipairs(trade) do
+            local trade_interesting = false
+            for output,amount in pairs(proposal[2]) do
+               if port.flag.ware_economy:needs(output) then
+                  trade_interesting = true
+                  break
+               end
+            end
+            if trade_interesting then
+               for input,amount in pairs(proposal[1]) do
+                  if port:get_wares(input) < amount then
+                     trade_interesting = false
+                     break
+                  end
+               end
+               if trade_interesting then
+                  for input,amount in pairs(proposal[1]) do
+                     port:set_wares(input, port:get_wares(input) - amount)
+                  end
+                  for output,amount in pairs(proposal[2]) do
+                     port:set_wares(output, port:get_wares(output) + amount)
+                  end
+                  sleep(3600)
+               end
+            end
+         end
+      end
+      sleep(7200)
+   end
+end
+
+function check_objective_wood()
+   local o = add_campaign_objective(obj_wood)
+   while not check_for_buildings(p1, {frisians_woodcutters_house = 3, frisians_foresters_house = 4}) do
+      sleep(4273)
+   end
+   set_objective_done(o)
+   campaign_message_box(secured_wood)
+end
+
+function check_objective_block()
+   local o = add_campaign_objective(obj_block)
+   while true do
+      sleep(4916)
+      local sites = 0
+      for x = 84, 94 do
+         for y = 114, 122 do
+            local f = map:get_field(x, y)
+            if f.immovable ~= nil and f.immovable.descr.type_name == "militarysite" and f.immovable.fields[1] == f and f.owner == p1 then
+               sites = sites + 1
+               if sites >= 3 then
+                  set_objective_done(o)
+                  campaign_message_box(secured_block)
+                  return
+               end
+            end
+         end
+      end
+   end
+end
+
+function check_objective_uplands()
+   local o = add_campaign_objective(obj_uplands)
+   while true do
+      sleep(4471)
+      local nr = 0
+      for _,descr in pairs(p1.tribe.buildings) do
+         for _,b in pairs(p1:get_buildings(descr.name)) do
+            if b.fields[1].height > 40 then
+               nr = nr + 1
+               if nr >= 15 then
+                  set_objective_done(o)
+                  campaign_message_box(secured_uplands)
+                  return
+               end
+            end
+         end
+      end
+   end
+end
+
+function mission_thread()
+   include "map:scripting/init_p2.lua"
+   run(check_fish)
+   campaign_message_box(intro_1)
+   campaign_message_box(intro_2)
+   campaign_message_box(intro_3)
+   campaign_message_box(intro_4)
+
+   scroll_to_field(map.player_slots[1].starting_field)
+   include "map:scripting/init_p1.lua"
+   p1:reveal_fields(map.player_slots[2].starting_field:region(9))
+   local reveal_fields = {}
```

Comment:
the reveal function takes much to long due to the enormous amount of fields to reveal. furthermore it does not seem very logic that the band of reebaud has instantaneous view of Helgolnds industry.    
what about the following sequennce:
1. place the helgoland industry and reduce it by 1 ship.  
2. instantenously hide all fields permanently. 
3. after message intro 3 place a ship near the HQ (in the harbor like structure)
4. reveal a concentric region around the the ship
5. run message Intro 4
6. after message intro 5 reveal everything that the helgolanders would have seen.
7. after intro 8 start to reveal the rest of the island randomy
8. afterwards start revealing the rest of the map. but we need to do this in small quadratic rectangulars in parralell.  
9. that might look good and would fit the story line.